File: algorithms/[1046]最后一块石头的重量/solution.ts
// topics = ["数组"]

/**
 * HEAP
 * time O(n * nlogn), space O(1), n 为 stones 长度
 */
function lastStoneWeight(stones: number[]): number {
  stones.sort((a, b) => a - b);

  while (stones.length > 1) {
    const d = stones.pop()! - stones.pop()!;
    if (d) {
      stones.push(d);
      stones.sort((a, b) => a - b);
    }
  }

  return stones[0] ?? 0;
}

File: algorithms/[1046]最后一块石头的重量/solution.py
# topics = ["堆"]

from typing import List
import heapq


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        """
        HEAP
        time O(nlogn), space O(n), n 为 stones 长度
        """
        hp: List[int] = []

        for stone in stones:
            heapq.heappush(hp, -1 * stone)

        while len(hp) > 1:
            x = heapq.heappop(hp)
            y = heapq.heappop(hp)
            if x < y:
                heapq.heappush(hp, x - y)

        return hp[0] * -1 if len(hp) else 0
